core/systems/project/_assembly.ex

```diff
@@ -41,13 +43,41 @@ defmodule Systems.Project.Assembly do
     |> Repo.transaction()
   end
 
+  @spec create(any(), any(), :empty) :: any()
   def create(name, user, :empty) do
-    project = prepare_project(name, [], user)
-
-    Multi.new()
-    |> Multi.insert(:project, project)
-    |> EctoHelper.run(:auth, &update_auth/2)
-    |> Repo.transaction()
+    project_changeset = prepare_project(name, [], user)
+
+    transaction_result =
+      Multi.new()
+      |> Multi.insert(:project, project_changeset)
+      |> EctoHelper.run(:auth, &update_auth/2)
+      |> Repo.transaction()
+
+    project =
+      case transaction_result do
+        {:ok, %{project: project}} ->
+          project
+
+        {:error, failed_operation, _failed_value, _changes} ->
+          Logger.warning("Transaction failed at #{failed_operation}")
+          nil
+      end
+
+    if project do
+      project = Repo.preload(project, :root)
+
+      changeset =
+        Storage.Public.prepare_endpoint(:builtin, %{key: "project_node=#{project.root.id}"})
+
+      if changeset.valid? do
+        create_item(changeset, dgettext("eyra-storage", "default.name"), project.root)
+      else
+        Logger.warning("Failed to prepare storage endpoint")
+        {:error, changeset}
+      end
+    else
+      {:error, "Project creation failed"}
+    end
```

Comment:
Looks good 👍 

Next thing is to make the "storage creation" part of the transaction to prevent any project to exist without storage:

 ```
   Multi.new()
    |> Multi.insert(:project, project)
    |> Multi.insert(:storage, fn %{project: inserted_project} -> prepare_storage(inserted_project) end))
    |> EctoHelper.run(:auth, &update_auth/2)
    |> Repo.transaction()
```

The idea is that you can capture the just inserted project from the transaction and use this one. Now the storage will not be created if the project failed to be created.

Reply:
Please check out the new version of this, and confirm this is what you were looking for.